llms/mixtral/convert.py

```diff
@@ -1,59 +1,152 @@
+    quantized_weights = {}
+    for k, v in all_weights.items():
+        if k not in weights:
+            continue
+        quantized_weights[k] = v
+        prefix = k.split(".")[:-1]
+        for qw in ["scales", "biases"]:
+            if (k := ".".join(prefix + [qw])) in all_weights:
+                quantized_weights[k] = all_weights[k]
+
+    # Update the config:
+    quantized_config["quantization"] = {
+        "group_size": args.q_group_size,
+        "bits": args.q_bits,
+    }
+    return quantized_weights, quantized_config
 
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser(description="Convert Mixtral weights to MLX.")
     parser.add_argument(
-        "--model-path",
+        "--torch-path",
+        type=str,
+        default="Mixtral-8x7B-v0.1",
+        help="The path to the PyTorch model.",
+    )
+    parser.add_argument(
+        "--mlx-path",
         type=str,
-        default="Mixtral-8x7B-v0.1/",
-        help="The path to the Mixtral model. The MLX model weights will also be saved there.",
+        default="mlx_model",
+        help="The path to save the MLX model.",
+    )
+    parser.add_argument(
+        "-q",
+        "--quantize",
+        help="Generate a quantized model.",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--q_group_size",
+        help="Group size for quantization.",
+        type=int,
+        default=64,
+    )
+    parser.add_argument(
+        "--q_bits",
+        help="Bits per weight for quantization.",
+        type=int,
+        default=4,
     )
     args = parser.parse_args()
-    model_path = Path(args.model_path)
+    torch_path = Path(args.torch_path)
+    mlx_path = Path(args.mlx_path)
+    mlx_path.mkdir(parents=True, exist_ok=True)
 
     with open("params.json") as fid:
-        args = json.load(fid)
-        args["model_type"] = "mixtral"
-    with open(model_path / "config.json", "w") as f:
-        json.dump(args, f, indent=4)
+        config = json.load(fid)
 
-    torch_files = glob.glob(str(model_path / "consolidated.*.pt"))
+    # Copy tokenizer
+    shutil.copyfile(
+        str(torch_path / "tokenizer.model"),
+        str(mlx_path / "tokenizer.model"),
+    )
+
+    # Convert and save model in shards
+    torch_files = glob.glob(str(torch_path / "consolidated.*.pt"))
     torch_files = sorted(torch_files, key=lambda tf: int(tf.split(".")[-2]))
     for e, tf in enumerate(torch_files):
         print(f"[INFO] Converting file {e + 1}/{len(torch_files)}")
-        state = torch.load(tf)
-        new_state = {}
-        for k, v in state.items():
-            new_state.update(convert(k, v, args))
-        np.savez(str(model_path / f"weights.{e}.npz"), **new_state)
+        weights = convert(tf, config)
+        if args.quantize:
+            print("[INFO] Quantizing")
+            weights, config = quantize(weights, config, args)
+        np.savez(str(mlx_path / f"weights.{e}.npz"), **weights)
```

Comment:
So currently using `mx.savez` without pre-evaluating would result in memory blowing up (bug 239 in mlx) but is there another reason to not use it? Did we encounter that before?

Should we change it to `mx.savez` after that is fixed?

Reply:
Actually the reason is simpler: `np.savez` handles `np` and `mx` arrays whereas `mx.savez` can only save `mx` arrays. So I left it at `np` rather than try to convert them to `mx` arrays in cases where no quantization is used.